nilearn/input_data/tests/test_nifti_masker.py

```diff
@@ -110,6 +111,39 @@ def test_mask_3d():
         assert_raises(TypeError, masker.fit)
 
 
+def test_mask_4d():
+    # Dummy mask
+    mask = np.zeros((10, 10, 10))
+    mask[3:7, 3:7, 3:7] = 1
+    mask_bool = mask.astype(bool)
+    mask_img = Nifti1Image(mask, np.eye(4))
+    n_mask_vox = mask_bool.sum()
+
+    # Dummy data
+    data = np.zeros((10, 10, 10, 3))
+    data[..., 0] = 1
+    data[..., 1] = 2
+    data[..., 2] = 3
+    data_img_4d = Nifti1Image(data, np.eye(4))
+    data_imgs = [Nifti1Image(data[..., 0], np.eye(4)),
+        Nifti1Image(data[..., 1], np.eye(4)),
+        Nifti1Image(data[..., 2], np.eye(4))]
+
+    # check whether transform is indeed selecting niimgs subset
+    sample_mask = (True, False, True)
+    masker = NiftiMasker(mask_img=mask_img, sample_mask=sample_mask)
+    masker.fit()
+    data_trans = masker.transform(data_imgs)
+    data_trans_direct = data_img_4d.get_data()[mask_bool, :][..., sample_mask]
+    data_trans_direct = np.swapaxes(data_trans_direct, 0, 1)
```

Comment:
Isn't it simpler to do (please use better names than mine):

```python

masker2 = NiftiMasker(mask_img=mask_img)
masker2.fit()
data_imgs_subset = index_img(data_imgs, sample_mask)
data_trans_direct = masker2.transform(data_imgs_subset)
```

Basically you want to make sure that using sample_mask in the masker is the same as using index_img on the data and then a masker without sample_mask.